### PR TITLE
Merge netsim mig to master

### PR DESCRIPTION
**Pull request by holisticode**
This PR is for the migration of all tests running a simulation in the `swarm/network/stream` package to the new `swarm/network/simulation` simulation framework.

It is also a major refactoring of these tests.

it obsoletes and removes `swarm/network/stream/testing` subpkg

### CODE REVIEW

File: swarm/network/stream/snapshot_sync_test.go [hunk 1]
```diff
@@ -138,230 +84,283 @@ func TestSyncing(t *testing.T) {
...
+			//default test
+			chnkCnt = []int{4, 32}
+			nodeCnt = []int{32, 16}
+		}
+		for _, chnk := range chnkCnt {
+			for _, n := range nodeCnt {
+				log.Info(fmt.Sprintf("Long running test with %d chunks and %d nodes...", chnk, n))
+				err := testSyncingViaDirectSubscribe(chnk, n)
+				if err != nil {
+					t.Fatal(err)
```
Comment (janos) [inline swarm/network/stream/snapshot_sync_test.go]: Just once in 200 iterations, I got this error:

```
--- FAIL: TestSyncingViaDirectSubscribe (61.21s)
        snapshot_sync_test.go:120: context deadline exceeded
``` 

This test usually passes in around 15 seconds, so having the deadline at 60s is safe enough. This means that TestSyncingViaDirectSubscribe will probably be flaky even with longer deadline, but the chance for failure is extremely low by my results. I see no easy way to reproduce the problem and to debug it, so maybe just to leave it as it is.

File: swarm/network/stream/snapshot_retrieval_test.go [hunk 2]
```diff
@@ -482,259 +257,134 @@ simulation's `action` function.
...
-			}
-			rpcSubscriptionsWg.Add(1)
-			go func() {
-				<-wdDoneC
-				rpcSubscriptionsWg.Done()
-			}()
-		}
+			r := NewRegistry(addr, delivery, db, state.NewInmemoryStore(), &RegistryOptions{
+				DoSync:          true,
+				SyncUpdateDelay: 0,
```
Comment (janos) [inline swarm/network/stream/snapshot_retrieval_test.go]: SyncUpdateDelay is time.Duration and if it is set to its default value of 0 (or not set at all), the delay will be set to 15s. If that is the intension, then this line can be removed, but if the intension was to lower the update delay, I am suggesting to set it to a reasonable low duration.

File: swarm/network/stream/snapshot_sync_test.go [hunk 3]
```diff
@@ -18,45 +18,33 @@ package stream
...
-	conf     *synctestConfig
-	ids      []discover.NodeID
-	datadirs map[discover.NodeID]string
-	ppmap    map[string]*network.PeerPot
-
-	live    bool
-	history bool
-
-	longrunning = flag.Bool("longrunning", false, "do run long-running tests")
+	//this global is needed to run the RPC test
```
Comment (janos) [inline swarm/network/stream/snapshot_sync_test.go]: Is there still an RPC test or it was removed? As globalConf is only related to testSyncingViaDirectSubscribe, maybe it should not be global anymore? In any case, the comment may be misleading.

File: swarm/network/stream/snapshot_sync_test.go [hunk 4]
```diff
@@ -138,230 +84,283 @@ func TestSyncing(t *testing.T) {
...
+		// File retrieval check is repeated until all uploaded files are retrieved from all nodes
+		// or until the timeout is reached.
+		allSuccess := false
+		var gDir string
+		var globalStore *mockdb.GlobalStore
+		if *useMockStore {
+			gDir, globalStore, err = createGlobalStore()
 			if err != nil {
-				return err
+				//return false, fmt.Errorf("Something went wrong; using mockStore enabled but globalStore is nil")
```
Comment (janos) [inline swarm/network/stream/snapshot_sync_test.go]: Remove commented code.

File: swarm/network/stream/snapshot_sync_test.go [hunk 5]
```diff
@@ -138,230 +84,283 @@ func TestSyncing(t *testing.T) {
...
+				os.RemoveAll(gDir)
+				err := globalStore.Close()
+				if err != nil {
+					log.Error("Error closing global store! %v", "err", err)
+				}
+			}()
+		}
+		for !allSuccess {
+			for _, id := range nodeIDs {
+				//log.Trace("file uploaded", "node", id, "key", key.String())
```
Comment (janos) [inline swarm/network/stream/snapshot_sync_test.go]: Remove commented code.

File: swarm/network/stream/snapshot_sync_test.go [hunk 6]
```diff
@@ -370,117 +369,90 @@ func runSyncTest(chunkCount int, nodeCount int, live bool, history bool) error {
...
-		for range timingTicker.C {
-			for i := 0; i < len(ids); i++ {
-				log.Trace(fmt.Sprintf("triggering step %d, id %s", i, ids[i]))
-				trigger <- ids[i]
+		// File retrieval check is repeated until all uploaded files are retrieved from all nodes
+		// or until the timeout is reached.
+		allSuccess := false
+		for !allSuccess {
+			for _, id := range nodeIDs {
+				//log.Trace("file uploaded", "node", id, "key", key.String())
```
Comment (janos) [inline swarm/network/stream/snapshot_sync_test.go]: Remove commented code.

File: swarm/network/stream/syncer_test.go [hunk 7]
```diff
@@ -46,222 +50,195 @@ func TestSyncerSimulation(t *testing.T) {
...
 	defer cancel()
 
-	// create simulation network with the config
-	sim, teardown, err := streamTesting.NewSimulation(conf)
-	var rpcSubscriptionsWg sync.WaitGroup
-	defer func() {
-		rpcSubscriptionsWg.Wait()
-		teardown()
-	}()
+	_, err := sim.AddNodesAndConnectFull(nodes)
```
Comment (janos) [inline swarm/network/stream/syncer_test.go]: Is full connectivity required there? Chain would be good enough.

File: swarm/network/stream/syncer_test.go [hunk 8]
```diff
@@ -46,222 +50,195 @@ func TestSyncerSimulation(t *testing.T) {
...
-		size := chunkCount * chunkSize
-		_, wait, err := rrFileStore.Store(ctx, io.LimitReader(crand.Reader, int64(size)), int64(size), false)
-		if err != nil {
-			t.Fatal(err.Error())
-		}
-		// need to wait cos we then immediately collect the relevant bin content
-		wait(ctx)
-		if err != nil {
-			t.Fatal(err.Error())
+		if *waitKademlia {
```
Comment (zelig) [inline swarm/network/stream/syncer_test.go]: ni suggest eliminating this variable/flag completely. 
It is either correct and necessary to wait or not, its not an option

File: swarm/network/stream/intervals_test.go [hunk 9]
```diff
@@ -22,52 +22,22 @@ import (
...
-	})
-
-	go func() {
-		waitPeerErrC <- waitForPeers(r, 1*time.Second, peerCount(id))
-	}()
-	return &TestExternalRegistry{r}, nil
-}
-
 func TestIntervals(t *testing.T) {
 	testIntervals(t, true, nil, false)
```
Comment (janos) [inline swarm/network/stream/intervals_test.go]: As this test only hangs on linux with Go 1.9, I am suggesting to temporary set this at start of TestIntervals function:

```
	if runtime.GOOS == "linux" && strings.HasPrefix(runtime.Version(), "go1.9") {
		t.Skip()
	}
```

File: swarm/network/stream/snapshot_retrieval_test.go [hunk 10]
```diff
@@ -199,261 +114,129 @@ simulation's `action` function.
...
-			log.Info("Uploading for history")
-			//If testing only history, we upload the chunk(s) first
-			conf.hashes, randomFiles, err = uploadFilesToNodes(nodes)
-			if err != nil {
-				return err
-			}
-		}
+			r := NewRegistry(addr, delivery, db, state.NewInmemoryStore(), &RegistryOptions{
+				DoSync:          true,
+				SyncUpdateDelay: 3 * time.Second,
```
[the remark this answers is not in the document]
Comment (holisticode) [reply swarm/network/stream/snapshot_retrieval_test.go]: If it is set to 0, the code will automatically initialize it to `15 * time.Second`. Thus I have this value for now.

**Pull request merged**